src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs

```diff
@@ -144,33 +145,59 @@ private bool ValidateTypeHierarchy(object item, Type itemType, ODataUrlValidatio
             validatedTypes.Add(itemType);
 
             // find any rules defined on an interface defined on the type of the item
-            foreach (Type interfaceType in itemType.GetInterfaces())
+            // GetInterfaces() returns derived interfaces, so only do this at the top level
+            if (includeInterfaces)
             {
-                if (validatedTypes.Add(interfaceType) && !this.unusedTypes.ContainsKey(interfaceType))
+                foreach (Type interfaceType in GetInterfaces(itemType))
                 {

```

Comment:
Can we use for loops as its array

Reply:
Good call; I like it.

---
In reply to: [518474457](https://github.com/OData/odata.net/pull/1917#discussion_r518474457) [](ancestors = 518474457)